src/inmanta/server/services/environment_metrics_service.py

```diff
@@ -0,0 +1,144 @@
+from typing import Dict, List
+
+from inmanta.data import EnvironmentMetricsCounter, EnvironmentMetricsNonCounter
+from inmanta.protocol.exceptions import ServerError
+from inmanta.server import SLICE_DATABASE, SLICE_ENVIRONMENT_METRICS, SLICE_TRANSPORT, protocol
+
+LOGGER = logging.getLogger(__name__)
+
+COLLECTION_INTERVAL_IN_SEC = 60
+
+
+class MetricType(str, Enum):
+    """
+    There are 3 types of metrics: metrics that represent a counter, metrics that doesn't represent a counter and
+    the compile_rate metric which is a special case. An examples of the count type metric is the agent_count metric which counts
+    the number of agents at a timestamp. An example of a non_count metric is the compile_time metric. This metric counts the
+    numer of compiles between the current timestamp and the last one and the total compile time.
+    The compile_rate metric is a special case as it is being calculated using only the count field of the compile_time metric.
+    """
+
+    COUNT = "count"
+    NON_COUNT = "non_count"
+    COMPILE_RATE = "compile_rate"
+
+
+class MetricsCollector(abc.ABC):
+    @abc.abstractmethod
+    def get_metric_name(self) -> str:
+        """
+        Returns the name of the metric collected by this MetricsCollector.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_metric_type(self) -> MetricType:
+        """
+        Returns the type of Metric collected by this metrics collector (count, non-count, etc.).
+        This information is required by the `EnvironmentMetricsService` to know how the data
+        should be aggregated.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    async def get_metric_value(self, start_interval: datetime, end_interval: datetime) -> dict[str, int]:
+        """
+        Invoked by the `EnvironmentMetricsService` at the end of the metrics collection interval.
+        Returns the metrics collected by this MetricCollector within the past metrics collection interval.
+
+        The *_interval arguments are present, because this method is intended to perform a query on
+        the database. No in-memory state is being stored by this metrics collector. The provided interval
+        should be interpreted as [start_interval, end_interval[
+
+        :param start_interval: The start time of the metrics collection interval (inclusive).
+        :param end_interval: The end time of the metrics collection interval (exclusive).
+        :result: The metrics collected by this MetricCollector within the past metrics collection interval.
+        """
+        raise NotImplementedError()
+
+
+class EnvironmentMetricsService(protocol.ServerSlice):
+    """Slice for the management of metrics"""
+
+    def __init__(self) -> None:
+        super(EnvironmentMetricsService, self).__init__(SLICE_ENVIRONMENT_METRICS)
+        self.metrics_collectors: Dict[str, MetricsCollector] = {}
+        self.previous_timestamp = datetime.now()
```

Comment:
this is not, strictly speaking, correct. 

the previous timestamp is the last one available in the DB. 

This may work, but it is tricky.

Reply:
I think this can work as well, we just have to be a bit careful